### ISSUE TITLE
'NoneType' object has no attribute 'segments'

### ISSUE THREAD

**Issue by hrsma2i** [sqlfluff/sqlfluff#3918]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The following SQL is not formatted about L042 for subqueries when using `union all` together and I got an error.

```sql
with
  b as (
    select
      x
      , y
    from (select
      x
      , y
      , z
      from a)
  )

select
  x
  , y
from a
union all
select
  x
  , y
from b
```

**Formatting correctly works like the following "Expected Behaviour" when I comment out the `union all` clause.**

### Expected Behaviour

It's expected to format the SQL like:

```sql
with prep_1 as (select
  x
  , y
  , z
  from a
)

, b as (
  select
    x
    , y
  from prep_1
)

select
  x
  , y
from a
union all
select
  x
  , y
from b
```

### Observed Behaviour

Formatting can't work and I got the following error:

```
==== finding fixable violations ====
CRITICAL   [L042] Applying rule L042 threw an Exception: 'NoneType' object has no attribute 'segments' 
Traceback (most recent call last):
  File "/Users/hiroshi.matsui/.pyenv/versions/3.9.6/lib/python3.9/site-packages/sqlfluff/core/rules/base.py", line 508, in crawl
    res = self._eval(context=context)
  File "/Users/hiroshi.matsui/.pyenv/versions/3.9.6/lib/python3.9/site-packages/sqlfluff/rules/L042.py", line 191, in _eval
    ctes.compose_select(
  File "/Users/hiroshi.matsui/.pyenv/versions/3.9.6/lib/python3.9/site-packages/sqlfluff/rules/L042.py", line 388, in compose_select
    from_clause_children = Segments(*from_clause.segments)
AttributeError: 'NoneType' object has no attribute 'segments'
==== no fixable linting violations found ====                                                      
All Finished 📜 🎉!
  [1 unfixable linting violations found]
```

### How to reproduce

```
$ sqlfluff fix sample.sql
```
sample.sql is the above SQL.

### Dialect

bigquery

### Version

sqlfluff, version 1.3.2

### Configuration

```toml
[sqlfluff]
dialect = bigquery
fix_even_unparsable = true
# 変数名のケースはそのまま
exclude_rules = L014

[sqlfluff:indentation]
indented_joins = true
indented_ctes = true

[sqlfluff:rules]
tab_space_size = 2
comma_style = leading

[sqlfluff:rules:L010]
capitalisation_policy = lower

[sqlfluff:rules:L016]
ignore_comment_lines = true
ignore_comment_clauses = false

[sqlfluff:rules:L030]
extended_capitalisation_policy = lower

[sqlfluff:rules:L040]
capitalisation_policy = lower

[sqlfluff:rules:L042]
forbid_subquery_in = both

[sqlfluff:rules:L063]
extended_capitalisation_policy = lower
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by hrsma2i** [comment]
Thank you.
